### PR TITLE
[REVIEW] JupyterCon Demo notebooks

### PR DESCRIPTION
**Pull request by Nanthini10**
CPU and GPU notebooks for jupyterCon demo.

### CODE REVIEW

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 1]
```diff
@@ -0,0 +1,1803 @@
...
+    "\n",
+    "def get_latest_mlflow_model(tracking_uri, model_id):\n",
+    "    client = MlflowClient(tracking_uri=tracking_uri, registry_uri=tracking_uri)\n",
+    "    model = client.get_registered_model(model_id)\n",
+    "    latest_model = model.latest_versions[0]\n",
+    "\n",
+    "    return f\"MLFLOW_TRACKING_URI={tracking_uri} mlflow models serve --no-conda -m models:/{model_id}/{latest_model.version} -p 56767\"\n",
+    "\n",
+    "## Custom callback, for additional flexibility, based on MLflowCallback\n",
+    "class RAPIDSMLflowCallback(object):\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:198]: This would be nice to document especially in slides - how does this differ from the standard MLFlow callback?

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 2]
```diff
@@ -0,0 +1,1803 @@
...
+    "            mlflow.set_tags(tags) "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Feature Engineering\n",
+    "\n",
+    "The following functions are defined to perform a few feature engineering tasks on the data. The feature_engineering function is called on the dataframe df, in this function we use cuML's [LabelEncoder](https://docs.rapids.ai/api/cuml/stable/api.html#cuml.preprocessing.LabelEncoder.LabelEncoder) and [TargetEncoder](https://docs.rapids.ai/api/cuml/stable/api.html#cuml.preprocessing.TargetEncoder) to encode the categorical columns. Arithmetic Combinations on the numerical columns using xfeat. The Arithmetic Operation performed here is addition on two columns at once. We specify the operator and r, which is used to indicate how many columns need to be combined.\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:260]: You explain this more later, but maybe mention explicitly here something like "We're just going to create a LOT of features without much manual optimization. This is part of the charm of automated feature engineering - the next steps will help make sure that only the most useful generated features are kept."

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 3]
```diff
@@ -0,0 +1,1803 @@
...
+    "    y_test = y_test.reset_index(drop=True)\n",
+    "    \n",
+    "    # Need to do this to ensure we are appropriately assigning the split values\n",
+    "    df_train[TARGET_COL] = y_train\n",
+    "    df_test[TARGET_COL] = y_test\n",
+    "    \n",
+    "    for col in CAT_COLS:\n",
+    "        out_col = f'{col}_TE'\n",
+    "        lbl_enc = LabelEncoder(handle_unknown='ignore')\n",
+    "        tar_enc = TargetEncoder(n_folds=5, smooth=TARGET_ENC_SMOOTH, split_method=TARGET_ENC_SPLIT)\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:292]: I'm really liking the simplification here since eliminating the CPU codepath... thank you! Looks great

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 4]
```diff
@@ -0,0 +1,1803 @@
...
+    "    return df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Feature Selection and Hyper parameter Optimisation\n",
+    "\n",
+    "Now that we have some new features, how do we know they are relevant for the task or represent anything meaningful? We use the feature selection process to do this. This helps in selection of a subset of features that are most informative. This helps in simplifying the problem and ensures that we aren't overloading the system with unimportant features.\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:322]: phrasing is a little reundant... maybe 
```suggestion
    "Now that we have some new features, how do we know they are relevant for the task or represent anything meaningful? We use the feature selection process to do this. This helps in selection of a subset of features that are most informative. This simplifies the machine learning problem and ensures that we aren't overloading the model with unimportant features.\n",
```

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 5]
```diff
@@ -0,0 +1,1803 @@
...
+    "\n",
+    "Optuna provides a way to choose a selector which accepts a Pipeline object from xfeat. You can see in the `feature_selection` function we define a `Pipeline` that takes in an Explorer and a Selection Algorithm (`ChiSquareKBest`). We pass this to an Optuna Study object, along with an Objective function. Chi squared tests are used to test the independence of two events.\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:324]: Could be in slides instead, but a brief explanation of what ChiSquareBest does would be nice

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 6]
```diff
@@ -0,0 +1,1803 @@
...
+    "\n",
+    "For Feature Selection, we aim to select features, which are highly dependent on the response. This way, we can get features that will best determine the outcome.\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:326]:
```suggestion
    "For Feature Selection, we aim to pick only those features that are most useful in predicting the response.\n",
```

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 7]
```diff
@@ -0,0 +1,1803 @@
...
+    "\n",
+    "#### Objective Function\n",
+    "The objective function will be the one we optimize in Optuna Study. Objective function tries out different values for the parameters that we are tuning and saving the results in `study.trials_dataframes()`.\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:329]:
```suggestion
    "The objective function will be the one we optimize in Optuna Study. Objective function evaluates the model for the various values for the parameters that we are tuning and saves the results in `study.trials_dataframes()`.\n",
```

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 8]
```diff
@@ -0,0 +1,1803 @@
...
+    "\n",
+    "Let's define the objective function for this HPO task by making use of the `train_and_eval()`. You can see that we simply choose a value for the parameters and call the train_and_eval method, making Optuna very easy to use in an existing workflow.\n",
+    "\n",
+    "The objective remains constant over different samplers, which are built-in options in Optuna to enable the selection of different sampling algorithms that optuna provides. Some of the available ones include - GridSampler, RandomSampler, TPESampler, etc. We'll use TPESampler for this demo, but feel free to try different samplers to notice the changes in performance. Tree-structured Parzen Estimator or TPE is based on independent sampling. It fits a Gaussian Mixture Model for each parameter in a trial `l(x)` and another GMM `g(x)` for other parameters. And it optimizes the `l(x)/g(x)` ratio to yield the best parameters.\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:333]: I would say something like "the objective function does not need to be changed" - remains constant sounds like it is `f(x; theta) = c` a constant

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 9]
```diff
@@ -0,0 +1,1803 @@
...
+    "\n",
+    "### HPO Trials and Study\n",
+    "Optuna uses [study](https://optuna.readthedocs.io/en/stable/reference/study.html) and [trials](https://optuna.readthedocs.io/en/stable/reference/trial.html) to keep track of the HPO experiments. Put simply, a trial is a single call of the objective function while a set of trials make up a study. We will pick the optimal performing trial from a study to get the best parameters that were used in that run."
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:336]: Replace optimal performing with "best performing" - we don't actually know that it's an optimum, just that it's the best we've tried.

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 10]
```diff
@@ -0,0 +1,1803 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_and_eval(df, penalty='l2', C=1.0, l1_ratio=None, fit_intercept=True, selector=None, return_model=False):\n",
+    "    # Splitting data and prepping for selector fit\n",
+    "    X_train,  X_test, y_train, y_test = train_test_split(df,\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:347]: How is the test set passed through here? We want to make sure that the test set held out in feature engineering (esp target encoding) is not used for training, but I don't see how that works here? Similarly, the inner eval loop seems to split the data differently? Maybe I'm just misinterpreting. though.

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 11]
```diff
@@ -0,0 +1,1803 @@
...
+    "for col in CAT_COLS:\n",
+    "    df_[col] = df_[col].astype(\"object\")\n",
+    "\n",
+    "with timed(\"Feature Engineering\"):\n",
+    "    df_feature_eng = feature_engineering(df_)\n",
+    "    df_feature_eng[TARGET_COL] = df_feature_eng[TARGET_COL].astype('float64')\n",
+    "    score = train_and_eval(df_feature_eng)\n",
+    "    print(\"After feature eng: \", score)\n",
+    "\n",
+    "df_feature_eng.head()"
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:914]: Maybe add a little commentary? Like - "You can see additive features with names ending in "_plus" generated by xfeat and note that our 20(??) column dataset has grown to XX(?) columns with all of the generated features."

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 12]
```diff
@@ -0,0 +1,1803 @@
...
+    "\n",
+    "f = optuna.visualization.plot_param_importances(study)\n",
+    "Image(f.to_image(format=\"png\", engine='kaleido'))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following is a slice plot to better understand the parameter relationships. We see how the change in the parameter affects the performance of the model."
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:1623]: Maybe a little more detailed commentary? "We can see that choosing the correct penalty and number of features is critical to model accuracy. Some parameters like l1_ratio had very little impact. Rather than spending our time tuning these manually, Optuna HPO has handled it for us."

**Review by JohnZed — CHANGES_REQUESTED** [review]
Looking really good overall! Feels simpler and easier to read than previous revs. I think we need to work out the train/test split topic noted below, but everything else is just tiny polishing commentary...

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb (hunk 1 above)
Comment (Nanthini10) [reply optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:198]: @drobison00 is getting the slides ready for this, and he will be talking about it during the walkthrough too.

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb (hunk 5 above)
Comment (Nanthini10) [reply optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:324]: I have it in the slides and also a few lines about the test is included in the next line.

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb (hunk 6 above)
Comment (Nanthini10) [reply optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:326]: This is actually the explanation for Chi-square test. That's how it performs feature selection, I'll word it better to make this clear.

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb (hunk 10 above)
Comment (Nanthini10) [reply optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:347]: I'm not sure I understand why the test set in Feature Engineering should not be used here? With feature engineering - I split the data and perform the encoding. Once the data is changed, I'm concatenating the dataframes to continue the rest of the workflow.

Which loop are you referring to? I've set the random_state for all splits to ensure that the results are reproducible with the splits.

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 13]
```diff
@@ -906,149 +925,76 @@
     "    df_[col] = df_[col].astype(\"object\")\n",
     "\n",
     "with timed(\"Feature Engineering\"):\n",
-    "    df_feature_eng = feature_engineering(df_)\n",
-    "    df_feature_eng[TARGET_COL] = df_feature_eng[TARGET_COL].astype('float64')\n",
-    "    score = train_and_eval(df_feature_eng)\n",
+    "    df_feature_eng_train, df_feature_eng_test = feature_engineering(df_train, df_test)\n",
+    "    df_feature_eng_train[TARGET_COL] = df_feature_eng_train[TARGET_COL].astype('float64')\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:929]: Why cast to float64? We should be all float32 right?

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 14]
```diff
@@ -319,48 +313,67 @@
...
+    "        ----------\n",
+    "        score: AUC score of the validation set\n",
+    "        classifier, signature: Returned only when return_model is set to True\n",
+    "    \"\"\"    \n",
     "    if selector:\n",
-    "        # For the selector, the label also needs to be in the DF\n",
-    "        X_train[TARGET_COL] = y_train\n",
-    "        X_test[TARGET_COL] = y_test\n",
+    "        # Getting the label column as it is dropped in the selector\n",
+    "        y_train = df_train[TARGET_COL]\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:365]: kind of funny that in one branch you do y first and in the next branch do X first

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb [hunk 15]
```diff
@@ -1722,34 +1282,34 @@
...
-    "    try:\n",
-    "        resp = requests.post(url=\"http://%s:%s/invocations\" % (host, port), data=json.dumps(data), headers=headers)\n",
-    "        print('Classification: %s' % (\"ON-Time\" if resp.text == \"[0.0]\" else \"LATE\"))\n",
-    "        break\n",
-    "    except Exception as e:\n",
-    "        errmsg = \"Caught exception attempting to call model endpoint: %s\" % e\n",
-    "        print(errmsg)\n",
-    "        print(\"... Sleeping ...\")\n",
-    "        time.sleep(20)"
+    "# host='localhost'\n",
```
Comment (JohnZed) [inline optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:1289]: Not sure why these are included but commented out?

**Review by JohnZed — COMMENTED** [review]
Train/test split improvement looks great!  The optimization history plot looks funny now, though. It seems like a flat line.. does it need more trials or another tweak?

File: optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb (hunk 13 above)
Comment (Nanthini10) [reply optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb:929]: It needs to be float but either of them works fine, I'll make it float32
